### DISCUSSION TITLE
Is it ever a good idea to add children to next/document's Head?

### DISCUSSION THREAD

**Discussion by TyMick** [vercel/next.js#13408] [Help] [answered]
I was trying to fix a [multiple viewport meta tags](https://github.com/zeit/next.js/issues/13230) issue caused by adding a `<meta>` tag as a child of the `Head` component in `_document.js`, and I was thinking about adding some deduplicating functionality to [`_document.tsx`](https://github.com/zeit/next.js/blob/canary/packages/next/pages/_document.tsx) that's similar to [what already exists in `next/head`](https://github.com/zeit/next.js/blob/canary/packages/next/next-server/lib/head.tsx#L52-L112). But then I realized that if I did that, user-entered `<head>` elements in `_document.js` would override matching user-entered `<head>` elements from `_app.js` or individual pages.

I can't think of any `<head>` elements one would add to `_document.js` that can't be added to `_app.js` with `next/head`. I can see why `_document.js` needs its own `<Head />` for rendering purposes, but should we discourage adding *any* children to this `Head` like we're currently discouraging adding `<title>` tags to `_document.js`?

https://github.com/vercel/next.js/blob/2828b01950f30e2790a870908255e6c28d96eebb/packages/next/pages/_document.tsx#L302-L306

**Comment by lfades** [comment, marked as answer]
`next/head` is dynamic, you can use it multiple times, update it, remove it, and your `head` tag will reflect those changes.

`_document.js` is only rendered once, in the server, and for your entire page, this is the one that generates the initial `html` tag that will be used for your page and it's not a place to save any tags that rely on dynamic data, or that only apply for some pages. A good use case for this one is for example to add a custom favicon link, or google analytics, something that you know only needs to be added once and has to be always there.

`_app.js` is similar in the way that it's also applied globally, but it does get rendered again every time you switch the current page, so it's also dynamic (which is never the case for `_document.js`)

**Reply by TyMick** [reply]
Thanks for the clarification! So I take it `<meta name="viewport" content="viewport-fit=cover" />` isn't a weird thing to add to `_document.js`, then, after all. Now to figure out how to stop it from being overridden by the [default head](https://github.com/vercel/next.js/blob/2828b01950f30e2790a870908255e6c28d96eebb/packages/next/next-server/lib/head.tsx#L11-L17) from `next/head`...

**Reply by timneutkens** [reply]
@tywmick viewport is handled by next/head so overriding it should be done using next/head. e.g. in _app.js instead of _document.

**Reply by zvictor** [reply]
> viewport is handled by next/head so overriding it should be done using next/head

If that's true, then shouldn't we add a warning as we do for `<title>`?
That's what @tywmick proposed as "option 1" in his comment at https://github.com/zeit/next.js/issues/13230#issuecomment-634119460.

Otherwise, people will keep trying to keep it in the `_document.js` since it's more performant there. Also, documenting this and @lfades' explanation above in some FAQ or "Best practices" document sounds like something needed by now. [It is indeed confusing for newcomers](https://github.com/zeit/next.js/discussions/12290).

**Reply by timneutkens** [reply]
For sure we can warn 👍

**Reply by TyMick** [reply]
Sounds good, I'll get on adding that warning! And maybe while I'm at it I could add to the [Custom <code>\`Document\`</code>](https://nextjs.org/docs/advanced-features/custom-document) page a list of `<head>` elements that should only be handled by `next/head`? @timneutkens, would those be the element types that `head.tsx`'s [`unique()`](https://github.com/zeit/next.js/blob/aef7f27072112d0d723c50c28ddba864397f6bfd/packages/next/next-server/lib/head.tsx#L50-L112) function checks to filter out duplicates? That is,

- `<title>`
- `<base>`
- `<meta name=...`
- `<meta httpEquiv=...`
- `<meta charSet=...`
- `<meta itemProp=...`

Come to think of it, would it be good to add warnings for *all* of these when added to `next/document`'s `<Head>`, or should I just add the `viewport` warning for now?

**Reply by timneutkens** [reply]
I think only viewport is fine for now

**Reply by willemmulder** [reply]
@timneutkens Hi Tim, what I do not understand is why the viewport meta should be handled by next/head and not in _document, especially because the viewport meta is a thing that many people want to set just once (i.e. sitewide). Then it would be more efficient to put it in _document, right?

**Reply by TyMick** [reply]
@willemmulder You're right, that does seem to violate the [breakdown](https://github.com/vercel/next.js/discussions/13408#discussioncomment-18763) @lfades gave at the beginning of this thread. 🤔 The only reason I can see right now for the current behavior is because `document/head` doesn't currently have any deduplicating functionality like `next/head`, so it isn't able to override the default viewport meta tag.

If `document/head` were given the ability to override just the default viewport, that would solve _one_ instance of multiple viewport tags (#13230) but allow for other instances of multiple viewport tags whenever a developer adds one to `document/head` _and_ to `next/head` somewhere else. So a warning is probably necessary _somewhere_, but come to think of it, maybe the latter situation the better one to warn about.

It would take some extra work, but if we let `document/head` override just the default viewport, and then adjusted `next/head` so that it can override the default but _can't_ override a viewport added to `document/head`, then Next.js could just warn people when they define multiple viewport tags themselves (in both `document/head` and `next/head`) instead of just warning them when they define a single viewport tag in a difficult place (`document/head`). A developer adding two of their own viewport tags does seem like the bigger blunder to me.

Now I'm intrigued. I'll open up a new issue (#19720) to see if anyone else thinks this is a good idea. 👍🏼

**Reply by willemmulder** [reply]
@tywmick Totally agree! I don't have a very deep understanding of the current process (and its design and implications) but would it even be possible to _allow_ people to override a viewport tag in `next/head`...? I.e.

1. `document/head` uses a) the defined viewport or b) else the default (if nothing is defined)
2. The result is passed to `next/head` which in turn will a) use the defined viewport if there is one or b) else will use the result from `document/head` (which will always be defined, right?)

Would that harm anything?

**Reply by TyMick** [reply]
@willemmulder Yeah, so currently, the default head lives in `next/head`, and it only includes two tags (or one if you're in AMP mode):

https://github.com/vercel/next.js/blob/c8cd77a856248346768042f4553a8513195311a4/packages/next/next-server/lib/head.tsx#L11-L17

Farther down in that file it [merges](https://github.com/vercel/next.js/blob/c8cd77a856248346768042f4553a8513195311a4/packages/next/next-server/lib/head.tsx#L114-L156) that default head with every other `next/head` child defined for a given page, filtering out any internal conflicts, and the result of that function is what gets injected into the HTML `head`. I feel like there's got to be a way, though, to tell `next/head` what's already been added to `document/head`, so `next/head` can remove anything that conflicts with it, including the defaults. Then, only if the conflict was something manually added to `next/head` by the user is there a need to warn them.

So the process would then be

1. `document/head` adds any viewport tag that's defined within it to the initial HTML, without concerning itself about defaults or anything. No warnings necessary.
2. The list of elements that `document/head` added gets passed to `next/head`'s filter somehow.
3. `next/head` filters out _all_ elements that conflict with `document/head` children.
   - If a conflict filtered out was part of `defaultHead`, that's perfect.
   - If a conflict filtered out was a manually defined `next/head` child, _then_ generate a warning, because the user goofed by defining conflicting tags in multiple places themselves.
4. `next/head` injects the results into the page.

**Reply by willemmulder** [reply]
Yes exactly!

**Comment by willemmulder** [comment]
Just for those that missed the link: issue https://github.com/vercel/next.js/discussions/19720 contains the proposal for a more permissive use of document/head!
